src/NovelStepTimer.cpp

```diff
@@ -0,0 +1,109 @@
+//
+// Created by tagoo on 12/09/2019.
+//
+
+// This is based on the StepTimer provided in the DirectX ToolKit
+// Original code is available under the MIT License
+
+#include "NovelStepTimer.h"
+#include <SDL2/SDL.h>
+
+namespace NovelRT {
+  StepTimer::StepTimer(uint32_t targetFrameRate) :
+    _frequency(SDL_GetPerformanceFrequency()),
+    _maxCounterDelta(_frequency / 10), // 1/10th of a second
+    _lastCounter(SDL_GetPerformanceCounter()),
+    _secondCounter(0),
+    _remainingTicks(0),
+    _elapsedTicks(0),
+    _totalTicks(0),
+    _targetElapsedTicks(TicksPerSecond / ((targetFrameRate != 0) ? targetFrameRate : 60)), // 60fps, if nothing else specified
+    _frameCount(0),
+    _framesPerSecond(0),
+    _framesThisSecond(0),
+    _isFixedTimeStep(targetFrameRate != 0) {
+  }
+
+  void StepTimer::resetElapsedTime() {
+    _lastCounter = SDL_GetPerformanceCounter();
+    _secondCounter = 0;
+    _remainingTicks = 0;
+    _framesPerSecond = 0;
+    _framesThisSecond = 0;
+  }
+
+  void StepTimer::tick(std::vector<NovelUpdateSubscriber> update) {
+    auto currentCounter = SDL_GetPerformanceCounter();
+    auto counterDelta = currentCounter - _lastCounter;
+
+    _lastCounter = currentCounter;
+    _secondCounter += counterDelta;
+
+    if (counterDelta > _maxCounterDelta)
+    {
+      // This handles excessibly large deltas to avoid overcompting.
+      // It is particularly beneficial to do this when debugging, for example
+      counterDelta = _maxCounterDelta;
+    }
+
+    // Convert to the "canonicalized" tick format of TicksPerSecond
+    // This will never overflow due to the clamping we did above
+    auto ticksDelta = (counterDelta * TicksPerSecond) / _frequency;
```

Comment:
Technically this can bug out if `_frequency` is greater than TicksPerSecond.

That can be fixed by converting to double first, but you have to adjust other logic as well (which isnt as straightforward to do since you have excessively small deltas that cant be converted to anything other than integer 0 (which leads to other logic getting "off").

Reply:
There are other workarounds as well, such as only dealing in terms of _frequency for your "ticks per second", but that also comes with tradeoffs.

At the end of the day, using 100ns ticks should be good enough as that is what most OS normalizes their High Performance Frequency Counters at and is significantly faster than most things can process frames in real world applications anyways.

(It also only causes bugs for variable refresh rate and can be easily solved by locking the refresh rate at some number such that 1 frame takes at least 100ns)